src/query_graph/path_tree.rs

```diff
@@ -31,17 +40,244 @@ pub(crate) struct PathTree<TTrigger: Eq + Hash, TEdge: Into<Option<EdgeIndex>>>
+        );
+        Self::from_paths(
+            graph,
+            node,
+            paths
+                .iter()
+                .map(|(path, selections)| (path.iter(), *selections))
+                .collect::<Vec<_>>(),
+        )
+    }
+}
+
+impl<TTrigger, TEdge> PathTree<TTrigger, TEdge>
+where
+    TTrigger: Eq + Hash,
+    TEdge: Copy + Hash + Eq + Into<Option<EdgeIndex>>,
+{
+    fn from_paths<'inputs>(
+        graph: Arc<QueryGraph>,
+        node: NodeIndex,
+        graph_paths_and_selections: Vec<(
+            impl Iterator<Item = GraphPathItem<'inputs, TTrigger, TEdge>>,
+            &'inputs Arc<NormalizedSelectionSet>,
+        )>,
+    ) -> Result<Self, FederationError>
+    where
+        TTrigger: 'inputs,
+        TEdge: 'inputs,
+    {
+        type MergeBy<'inputs, TEdge, TTrigger> = (TEdge, &'inputs Arc<TTrigger>);
+
+        /// `GraphPathIter` is a partly-consumed
+        /// `impl Iterator<Item = GraphPathItem<'_, TTrigger, TEdge>>`
+        struct Merged<'inputs, GraphPathIter> {
+            target_node: NodeIndex,
+            conditions: Option<Arc<OpPathTree>>,
+            sub_paths_and_selections: Vec<(GraphPathIter, &'inputs Arc<NormalizedSelectionSet>)>,
+        }
+
+        let mut merged = IndexMap::<MergeBy<_, _>, Merged<_>>::new();
+        let mut local_selection_sets = Vec::new();
+
+        for (mut graph_path_iter, selection) in graph_paths_and_selections {
+            let Some((generic_edge, trigger, conditions)) = graph_path_iter.next() else {
+                // End of an input `GraphPath`
+                local_selection_sets.push(selection.clone());
+                continue;
+            };
+            match merged.entry((generic_edge, trigger)) {
+                Entry::Occupied(entry) => {
+                    let existing = entry.into_mut();
+                    existing.conditions = merge_conditions(&existing.conditions, conditions);
+                    existing
+                        .sub_paths_and_selections
+                        .push((graph_path_iter, selection))
+                    // Note that as we merge, we don't create a new child
+                }
+                Entry::Vacant(entry) => {
+                    entry.insert(Merged {
+                        target_node: if let Some(edge) = generic_edge.into() {
+                            let (_source, target) = graph.edge_endpoints(edge)?;
+                            target
+                        } else {
+                            // For a "None" edge, stay on the same node
+                            node
+                        },
+                        conditions: conditions.clone(),
+                        sub_paths_and_selections: vec![(graph_path_iter, selection)],
+                    });
+                }
+            }
+        }
+
+        let childs = merged
+            .into_iter()
+            .map(|(key, value)| {
```

Comment:
The ordering of `Vec` items here will end up being different from the JS code.

The JS code groups items with the same edge (but different triggers) together, in the order in which the triggers were encountered. The JS code also records the order in which edges were encountered for the first time, and orders these same-edge groups according to that.

Reply:
I’ve done that but with an intermediate struct. Please take a look.